src/java/org/apache/cassandra/net/ResponseVerbHandler.java

```diff
@@ -56,6 +65,50 @@ public void doVerb(Message message)
         {
             MessagingService.instance().latencySubscribers.maybeAdd(cb, message.from(), latencyNanos, NANOSECONDS);
             cb.onResponse(message);
+            trackReplicaSensors(callbackInfo, message);
+        }
+    }
+
+    private void trackReplicaSensors(RequestCallbacks.CallbackInfo callbackInfo, Message message)
+    {
+        RequestSensors sensors = callbackInfo.callback.getRequestSensors();
+        // Eventhough RequestCallback.getRequestSensors() returns a NoOpRequestSensors instance by default, callbacks
+        // that override this method may have been initialized from a code path where sensors are not set.
+        if (sensors == null)
+            return;
+
+        if (callbackInfo instanceof RequestCallbacks.WriteCallbackInfo)
```

Comment:
Ideally, rather than retrieving the sensors and then populating them here, which requires an `instanceof`, we should add a `populateRequestSensors()` method directly to the callback and have specific implementations in there.

That said, AFAIU you've made this because we only have access to the `mutation` in `WriteCallbackInfo`, where:
```
this.mutation = shouldHint(allowHints, message, consistencyLevel) ? message.payload : null;
```
So the `mutation` could actually be `null`.

Two options here:
1. Make sure the `mutation` is never null in `WriteCallbackInfo`.
2. Rework things so that the `mutation` can be accessed from the callback.

Thoughts?

Reply:
I think you convinced me to give up and leave things as they are :)